### PR TITLE
Bug 1857743: Port stuck open when ep deleted before svc in unidling mode

### PR DESCRIPTION
**Pull request by tssurya**
When the ep is deleted before the svc in unidling proxy mode, the port
does not get deleted and remains stuck. This patch adds a new variable
pendingDeletion to hybrid proxier which keeps track of the svc if the
ep is the one that gets deleted first.

Signed-off-by: Surya Seetharaman <suryaseetharaman.9@gmail.com>

### CODE REVIEW

**Comment by tssurya** [comment]
/assign @squeed 

@squeed : I know this is not exactly what you had in mind, but I just wanted to check and see if this would work or not.

PS: I know I'd have to add tests at some point.

File: pkg/network/proxyimpl/hybrid/proxy.go [hunk 1]
```diff
@@ -153,9 +158,21 @@ func (p *HybridProxier) OnServiceDelete(service *corev1.Service) {
...
 	defer p.switchedToUserspaceLock.Unlock()
 
-	if isUsingUserspace, ok := p.usingUserspace[svcName]; ok && isUsingUserspace {
+	isUsingUserspace, ok := p.usingUserspace[svcName]
+	_, isPendingDeletion := p.pendingDeletion[svcName]
+
+	if (ok && isUsingUserspace) || isPendingDeletion {
 		klog.V(6).Infof("hybrid proxy: del svc %s/%s in unidling proxy", service.Namespace, service.Name)
 		p.unidlingProxy.OnServiceDelete(service)
+		if isPendingDeletion {
```
Comment (tssurya) [inline pkg/network/proxyimpl/hybrid/proxy.go:167]: I guess I could chalk this out to a separate function since the code is duplicated here and below.

**Comment by squeed** [comment]
I had a bit of a different fix in mind:

In both `OnServiceDelete` and `OnEndpointsDelete`, we put an entry in `pendingDeletions`. If the entry is already there, then we delete the values in `usingUserspace` and `switchedService`. Then we don't have to change the logic anywhere else.

**Comment by tssurya** [comment]
> I had a bit of a different fix in mind:
> 
> In both `OnServiceDelete` and `OnEndpointsDelete`, we put an entry in `pendingDeletions`. 

Q: but you mean only in the `unidlingproxy` mode right ? 

> If the entry is already there, then we delete the values in `usingUserspace` and `switchedService`. Then we don't have to change the logic anywhere else.

hmm yea I wanted to do it the way you said but I got a bit confused on the local cache deletion parts since `OnServiceDelete` uses the `switchedService` map and  `OnEndpointsDelete` uses the `usingUserspace` map and then got confused. But... wait! ah I think I get it now. So what you meant was:

- when we go to `OnServiceDelete` we add  an entry in `pendingDeletions[svcName]`. 
- when we go to `OnEndpointsDelete` we check if the entry is already present and if yes we delete it from both caches since that would clear endpoints and svc. It doesn't really matter even if we deleted both caches which is what is expected.
- same for viceversa.


What I basically do now, is fix what was broken which is calling the svc deletion in  `OnServiceDelete` in `unidlingproxy` mode if the endpoint got deleted already. I think this should also work since technically all this `unidlingproxy.deletion `takes place only when you idle and unidle. But anyhow I guess I'd have to test this properly.

**Comment by squeed** [comment]
> * when we go to `OnServiceDelete` we add  an entry in `pendingDeletions[svcName]`.
> 
> * when we go to `OnEndpointsDelete` we check if the entry is already present and if yes we delete it from both caches since that would clear endpoints and svc. It doesn't really matter even if we deleted both caches which is what is expected.
> 
> * same for viceversa.

Exactly. I'd just write a `cleanupState(nsName)` function that takes care of everything, then call it in `OnServiceDelete` and `OnEndpointsDelete`. That way we don't have to worry about ordering at all.

**Comment by tssurya** [comment]
> Exactly. I'd just write a `cleanupState(nsName)` function that takes care of everything, then call it in `OnServiceDelete` and `OnEndpointsDelete`. That way we don't have to worry about ordering at all.

cool, I've tried it your way and also err I am not really sure how/where to add the tests for this bug.

File: pkg/network/proxyimpl/hybrid/proxy.go [hunk 2]
```diff
@@ -140,6 +145,19 @@ func (p *HybridProxier) OnServiceUpdate(oldService, service *corev1.Service) {
 	}
 }
 
+func (p *HybridProxier) CleanupState(svcName types.NamespacedName) {
```
Comment (squeed) [inline pkg/network/proxyimpl/hybrid/proxy.go:148]: this should be a private function, so lowercase this name.

Comment (squeed) [reply pkg/network/proxyimpl/hybrid/proxy.go:148]: also, add a short docblock explaining what this does (handle endpoint and service deletions in any order)

**Comment by squeed** [comment]
Two very minor things, then this is good to go.

File: pkg/network/proxyimpl/hybrid/proxy.go (hunk 2 above)
Comment (tssurya) [reply pkg/network/proxyimpl/hybrid/proxy.go:148]: > also, add a short docblock explaining what this does (handle endpoint and service deletions in any order)

ack yea good point, I always forget to do this.

**Comment by tssurya** [comment]
CI failing due to:
`Waiting for container artifacts in pod release-latest in namespace`

This failure is being investigated by the testplatform team.

**Pull request merged**